File: SWEA/battle_ssafy/5105.py
from collections import deque


def solve():
    n = int(input())
    grid = []
    start_y, start_x = -1, -1

    for i in range(n):
        row = input()
        grid.append(row)
        if "2" in row:
            start_y, start_x = i, row.index("2")

    q = deque([(start_y, start_x, 0)])
    visited = [[False] * n for _ in range(n)]
    visited[start_y][start_x] = True
    DIR = ((-1, 0), (1, 0), (0, -1), (0, 1))
    while q:
        y, x, dist = q.popleft()
        for dy, dx in DIR:
            ny, nx = y + dy, x + dx
            if 0 <= ny < n and 0 <= nx < n and not visited[ny][nx]:
                if grid[ny][nx] == "3":
                    return dist
                if grid[ny][nx] != "1":
                    visited[ny][nx] = True
                    q.append((ny, nx, dist + 1))
    return 0


def main():
    for tc in range(1, int(input()) + 1):
        ans = solve()
        print(f"#{tc} {ans}")


if __name__ == "__main__":
    main()

File: SWEA/battle_ssafy/5105.cpp
#include <iostream>
#include <string>
#include <cstring>
#include <queue>

using namespace std;

int n, ans;
string map_[51];
int dist_[51][51];

struct Point
{
    int y, x;
    Point operator+(Point other) const
    {
        return {y + other.y, x + other.x};
    }
    bool isValid() const
    {
        return 0 <= y && y < n && 0 <= x && x < n && map() != '1' && dist() == -1;
    }
    char map() const
    {
        return map_[y][x];
    }
    int &dist() const
    {
        return dist_[y][x];
    }
} start;
constexpr Point DIR[4] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

int bfs()
{
    queue<Point> q;
    q.push(start);
    start.dist() = 0;
    while (!q.empty())
    {
        Point cur = q.front();
        q.pop();
        for (Point d : DIR)
        {
            Point nxt = cur + d;
            if (!nxt.isValid())
                continue;
            if (nxt.map() == '3')
                return cur.dist();
            nxt.dist() = cur.dist() + 1;
            q.push(nxt);
        }
    }
    return 0;
}

void init()
{
    memset(dist_, -1, sizeof(dist_));
}

void input()
{
    cin >> n;
    for (int i = 0; i < n; ++i)
    {
        cin >> map_[i];
        size_t j = map_[i].find('2');
        if (j != string::npos)
            start.y = i, start.x = j;
    }
}

void solve()
{
    ans = bfs();
}

void output(int testCase)
{
    cout << '#' << testCase << ' ' << ans << '\n';
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    for (int testCase = 1; testCase <= T; ++testCase)
    {
        init();
        input();
        solve();
        output(testCase);
    }

    return 0;
}